tests/functional/watcher_controller_test.go

```diff
@@ -371,4 +372,31 @@ var _ = Describe("Watcher controller", func() {
 		})
 	})
 
+	When("Watcher CR is created without container images defined", func() {
+		BeforeEach(func() {
+			DeferCleanup(th.DeleteInstance, CreateWatcher(watcherTest.Instance, GetDefaultWatcherSpec()))
+			DeferCleanup(
+				mariadb.DeleteDBService,
+				mariadb.CreateDBService(
+					watcherTest.Instance.Namespace,
+					GetWatcher(watcherTest.Instance).Spec.DatabaseInstance,
+					corev1.ServiceSpec{
+						Ports: []corev1.ServicePort{{Port: 3306}},
+					},
+				),
+			)
+			spec := GetDefaultWatcherSpec()
+			// This watcher is created without any container image is specified in
+			// the request
+			DeferCleanup(th.DeleteInstance, CreateWatcher(watcherTest.Instance, spec))
+
+		})
+		It("It has the expected container image defaults", func() {
+			Watcher := GetWatcher(watcherTest.Instance)
+			Expect(Watcher.Spec.APIContainerImageURL).To(Equal(util.GetEnvVar("RELATED_IMAGE_WATCHER_API_IMAGE_URL_DEFAULT", watcherv1beta1.WatcherAPIContainerImage)))
+			Expect(Watcher.Spec.DecisionEngineContainerImageURL).To(Equal(util.GetEnvVar("RELATED_IMAGE_WATCHER_DECISION_ENGINE_IMAGE_URL_DEFAULT", watcherv1beta1.WatcherDecisionEngineContainerImage)))
+			Expect(Watcher.Spec.ApplierContainerImageURL).To(Equal(util.GetEnvVar("RELATED_IMAGE_WATCHER_APPLIER_IMAGE_URL_DEFAULT", watcherv1beta1.WatcherApplierContainerImage)))
```

Comment:
IMO we should run three different cases to fully cover the defaulting functionality:

1. No container images defined and no env variable defined. Spec values should == watcherv1beta1.WatcherApplierContainerImage
2. No container images defined ane env variable defined to some fake values (this will require a different When case). Spec values should == util.GetEnvVar("RELATED_IMAGE_WATCHER_APPLIER_IMAGE_URL_DEFAULT")
3. When Watcher is created with container images values in the spec, Spec values should == to the defined ones (This will also required a different When case to set a different spec).

Reply:
@amoralej I need help on 3rd option, I am not sure i have done it correctly. https://github.com/raukadah/watcher-operator/blob/webhook_default_images/tests/functional/watcher_controller_test.go#L70 Please have a look.

Reply:
I'd say that's incorrect as you are modifying the spec after the reconcile loop is finised. I think you need to:

1. Add a new spec in addition to MinimalWatcherSpec with your fake values for the images. Something like:

https://github.com/raukadah/watcher-operator/blob/webhook_default_images/tests/functional/watcher_controller_test.go#L20-L24
 
2. Define an entirely new test watcher deployment using something similar to:

https://github.com/raukadah/watcher-operator/blob/webhook_default_images/tests/functional/watcher_controller_test.go#L27-L31

but instead of using MinimalWatcherSpec you need to use yours.

3. Check the values as in:

https://github.com/raukadah/watcher-operator/blob/webhook_default_images/tests/functional/watcher_controller_test.go#L32-L40

Note all this has to go into a new When block after the existing one, i.e. after 

https://github.com/raukadah/watcher-operator/blob/webhook_default_images/tests/functional/watcher_controller_test.go#L399

Reply:
Actually, Case 2 (getting defaults from env values) also need some rework from https://github.com/raukadah/watcher-operator/blob/webhook_default_images/tests/functional/watcher_controller_test.go#L63-L68 , but I'm not sure how to inject env values for a specific test in envtest. Unless we find a way we may need to skip that case from functional test coverage.

Reply:
Thank you @amoralej for the detailed writeup. Here is what I have implemented for testcase 3: https://github.com/raukadah/watcher-operator/blob/webhook_default_images/tests/functional/watcher_controller_test.go#L413

Reply:
For second case, I tried to go via this https://github.com/raukadah/watcher-operator/blob/webhook_default_images/tests/functional/watcher_controller_test.go#L390 Not sure how much it is correct.

Reply:
I think you are going in the right path, but i'm adding some suggestions.